src/PrestaShopBundle/Form/Admin/Sell/Product/ExtraModulesType.php

```diff
@@ -0,0 +1,135 @@
+
+namespace PrestaShopBundle\Form\Admin\Sell\Product;
+
+use PrestaShop\PrestaShop\Adapter\Module\ModuleDataProvider;
+use PrestaShop\PrestaShop\Core\Hook\HookDispatcherInterface;
+use PrestaShop\PrestaShop\Core\Module\ModuleRepository;
+use PrestaShopBundle\Form\Admin\Type\TranslatorAwareType;
+use Symfony\Component\Form\FormInterface;
+use Symfony\Component\Form\FormView;
+use Symfony\Component\OptionsResolver\OptionsResolver;
+use Symfony\Component\Translation\TranslatorInterface;
+
+/**
+ * This form type is used to display modules in an extra tab that regroup the modules implementing the displayAdminProductsExtra
+ * hook. This is not the recommended way to integrate in the product page anymore but we keep it for backward compatibility.
+ */
+class ExtraModulesType extends TranslatorAwareType
+{
+    public const HOOK_NAME = 'displayAdminProductsExtra';
+
+    /**
+     * @var HookDispatcherInterface
+     */
+    private $hookDispatcher;
+
+    /**
+     * @var ModuleDataProvider
+     */
+    private $moduleDataProvider;
+
+    /**
+     * @var ModuleRepository
+     */
+    private $moduleRepository;
+
+    public function __construct(
+        TranslatorInterface $translator,
+        array $locales,
+        HookDispatcherInterface $hookDispatcher,
+        ModuleDataProvider $moduleDataProvider,
+        ModuleRepository $moduleRepository
+    ) {
+        parent::__construct($translator, $locales);
+        $this->hookDispatcher = $hookDispatcher;
+        $this->moduleDataProvider = $moduleDataProvider;
+        $this->moduleRepository = $moduleRepository;
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    public function buildView(FormView $view, FormInterface $form, array $options)
+    {
+        parent::buildView($view, $form, $options);
+
+        $extraModules = $this->renderHooksArray($options['product_id']);
+        $view->vars['extraModules'] = $extraModules;
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    public function configureOptions(OptionsResolver $resolver)
+    {
+        parent::configureOptions($resolver);
+        $resolver
+            ->setDefaults([
+                'label' => $this->trans('Modules', 'Admin.Catalog.Feature'),
+                'form_theme' => '@PrestaShop/Admin/Sell/Catalog/Product/FormTheme/extra_modules.html.twig',
+                'required' => false,
+            ])
+            ->setRequired('product_id')
+            ->setAllowedTypes('product_id', 'int')
+        ;
+    }
+
+    /**
+     * This method is here to simulate the previous call of renderhooksarray('displayAdminProductsExtra', { 'id_product': productId })
+     * so it is basically a copy (slightly adapted) of the code from HookExtension::renderHooksArray method.
+     */
+    protected function renderHooksArray(int $productId): array
+    {
+        // The call to the render of the hooks is encapsulated into a ob management to avoid any call of echo from the
+        // modules.
+        ob_start();
```

Comment:
If 'ob' is a default security to avoid modules unmanaged output, shouldn't it be integrated directly to `dispatchRenderingWithParameters` method ? (and maybe be tested as a feature ? :thinking: )

Reply:
Maybe but I believe it would be a kind of breaking change 🤔 Besides as I explained in the comment this code is mostly copied from this existing code https://github.com/PrestaShop/PrestaShop/blob/2968d8370381cab170ea0470f0d704c07239fffc/src/PrestaShopBundle/Twig/HookExtension.php#L122

The purpose of this PR is not to refacto how we handle hook rendering but to make the product form more extendable by providing both generic and specific component This one is really specific to the page that's why I didn't put too much effort into refactoring it

Reply:
Then what do you think about creating an issue describing this `ob` use case and the one you mentioned to start a discussion / thinking about a potential refactoring ?
This way we could add in the code a "TODO" comment linked to this issue without too deep refactoring and we won't lose this refactoring task subject as well ^^